iot/api-client/end_to_end_example/cloudiot_pubsub_example_server.py

```diff
@@ -0,0 +1,211 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+r"""Sample server that pushes configuration to Google Cloud IoT devices.
+
+This example represents a server that consumes telemetry data from multiple
+Cloud IoT devices. The devices report telemetry data, which the server consumes
+from a Cloud Pub/Sub topic. The server then decides whether to turn on or off
+individual devices fans.
+
+This example requires the Google Cloud Pub/Sub client library. Install it with
+
+  $ pip install --upgrade google-cloud-pubsub
+
+If you are running this example from a Compute Engine VM, you will have to
+enable the Cloud Pub/Sub API for your project, which you can do from the Cloud
+Console. Create a pubsub topic, for example
+projects/my-project-id/topics/my-topic-name, and a subscription, for example
+projects/my-project-id/subscriptions/my-topic-subscription.
+
+You can then run the example with
+
+  $ python cloudiot_pubsub_example_server.py \
+    --project_id=my-project-id \
+    --pubsub_subscription=my-topic-subscription \
+    --api_key=YOUR_API_KEY
+"""
+
+import argparse
+import base64
+import json
+import sys
+from threading import Lock
+import time
+
+from googleapiclient import discovery
+from googleapiclient.errors import HttpError
+from oauth2client.service_account import ServiceAccountCredentials
+from google.cloud import pubsub_v1
+
+
+API_SCOPES = ['https://www.googleapis.com/auth/cloud-platform']
+API_VERSION = 'v1beta1'
+DISCOVERY_API = 'https://cloudiot.googleapis.com/$discovery/rest'
+SERVICE_NAME = 'cloudiot'
+
+
+def discovery_url(api_key):
+  """Construct the discovery url for the given api key."""
+  return '{}?version={}&key={}'.format(DISCOVERY_API, API_VERSION, api_key)
+
+
+class Server(object):
+  """Represents the state of the server."""
+
+  def __init__(self, service_account_json, api_key):
+    credentials = ServiceAccountCredentials.from_json_keyfile_name(
+        service_account_json, API_SCOPES)
+    if not credentials:
+      sys.exit('Could not load service account credential from {}'.format(
+          service_account_json))
+
+    self._service = discovery.build(
+        SERVICE_NAME,
+        API_VERSION,
+        discoveryServiceUrl=discovery_url(api_key),
+        credentials=credentials)
+
+    # Used to serialize the calls to the
+    # modifyCloudToDeviceConfig REST method. This is needed
+    # because the google-api-python-client library is built on top
+    # of the httplib2 library, which is not thread-safe. For more
+    # details, see: https://developers.google.com/
+    #     api-client-library/python/guide/thread_safety
+    self._update_config_mutex = Lock()
+
+  def _update_device_config(self, project_id, region, registry_id, device_id,
+                            data):
+    """Push the data to the given device as configuration."""
+    config_data = None
+    print 'The device ({}) has a temperature of: {}'.format(device_id,
```

Comment:
We generally avoid hanging indents like this

Reply:
Can you be more specific? Unless it's been fixed.